# Phase 2: Data Preprocessing & Leakage-Free Feature Pipeline

**Project**: Machine Learning Based Campaign Response Prediction for Bank Marketing  
**Phase**: 02 – Data Engineering & Preprocessing  
**Owner**: Athukorala M. B. (IT24101937)  
**Governance & Standards**: Aligned with Constitution Principles I–VI, [ADR-001 (Validation Split Strategy)](../docs/decision_log.md#adr-001), [ADR-002 (Categorical Encoding Strategy for Education)](../docs/decision_log.md#adr-002), and [ADR-004 (Dual Age Representation)](../docs/decision_log.md#adr-004)

### Purpose
This notebook implements the reproducible, leakage-free data engineering and feature preprocessing pipeline for the UCI Bank Marketing dataset (`bank-additional-full.csv`). It resolves the validation split strategy discrepancy via an empirical investigation (ADR-001), rectifies the encoding of customer educational attainment by replacing arbitrary ordinal ranking with one-hot encoding (ADR-002), establishes strict fit-on-train data isolation, applies modular Scikit-Learn transformers, and exports serialized artifacts for downstream model training.

### Inputs
- Raw Dataset: `data/raw/bank-additional-full.csv` (41,188 records, 20 features + target `y`)
- Modular Transformer: `src.transformers.BankFeatureEngineer`

### Outputs
- `data/processed/X_train_processed.csv`: Transformed training feature matrix (32,950 × 71)
- `data/processed/X_test_processed.csv`: Transformed holdout test feature matrix (8,238 × 71)
- `data/processed/y_train.csv`: Training target vector (32,950 records)
- `data/processed/y_test.csv`: Holdout target vector (8,238 records)
- `data/processed/preprocessor_pipeline.joblib`: Serialized fitted `Pipeline`
- `data/processed/feature_metadata.json`: Feature catalog and pipeline specifications (71 features)

In [1]:
import os
import sys
import json
from pathlib import Path

# Deterministically resolve project root independent of caller working directory
current_dir = Path.cwd().resolve()
if (current_dir / "src" / "transformers.py").exists():
    PROJECT_ROOT = current_dir
elif (current_dir.parent / "src" / "transformers.py").exists():
    PROJECT_ROOT = current_dir.parent
else:
    PROJECT_ROOT = next(
        (p for p in [current_dir] + list(current_dir.parents) if (p / "src" / "transformers.py").exists()),
        current_dir
    )

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import joblib

from sklearn.base import BaseEstimator, TransformerMixin
from sklearn.model_selection import train_test_split
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline
from sklearn.preprocessing import (
    OneHotEncoder,
    OrdinalEncoder,
    StandardScaler,
    RobustScaler
)

# Plotting & display configuration
sns.set_theme(style="whitegrid", palette="muted")
plt.rcParams.update({'figure.autolayout': True, 'font.sans-serif': 'Arial'})
pd.set_option('display.max_columns', None)
pd.set_option('display.width', 1000)

RANDOM_STATE = 42
np.random.seed(RANDOM_STATE)

print(f"Project root resolved to: {PROJECT_ROOT}")
print(f"NumPy version:        {np.__version__}")
print(f"Pandas version:       {pd.__version__}")
import sklearn
print(f"Scikit-Learn version: {sklearn.__version__}")
print(f"Joblib version:       {joblib.__version__}")
print("Environment initialized successfully.")

Project root resolved to: /Users/moni/Desktop/ML proj
NumPy version:        2.5.3
Pandas version:       3.0.6
Scikit-Learn version: 1.9.1
Joblib version:       1.6.0
Environment initialized successfully.


## 1. Raw Data Ingestion, Deduplication (ADR-003), & Target Isolation (PRE-01)

In strict adherence to **Constitution Principle I (*Reproducibility First*)**, the raw dataset located in `data/raw/bank-additional-full.csv` is treated as strictly read-only.

In accordance with **[ADR-003](file:///Users/moni/Desktop/ML%20proj/docs/decision_log.md#adr-003)** and **[EDA-17](file:///Users/moni/Desktop/ML%20proj/docs/eda_insight_log.md)**, exact duplicate rows are audited and pruned prior to splitting (`df_raw.drop_duplicates()`), preserving exactly **41,176 distinct customer records** and eliminating cross-split duplicate leakage.

In accordance with **PRE-01** and **Constitution Principle III (*No Data Leakage*)**, the target variable `y` is encoded as binary (0/1), while call `duration` is quarantined and permanently removed from the feature matrix because it is physically unobservable before placing a call.

In [2]:
# Locate dataset dynamically using PROJECT_ROOT anchor
candidates = [
    PROJECT_ROOT / "data" / "raw" / "bank-additional-full.csv",
    Path("../data/raw/bank-additional-full.csv"),
    Path("data/raw/bank-additional-full.csv"),
    Path("bank-additional-full.csv"),
    Path("../../data/raw/bank-additional-full.csv")
]

raw_data_path = None
for p in candidates:
    if p.exists():
        raw_data_path = p.resolve()
        break

if raw_data_path is None:
    raise FileNotFoundError("Could not locate 'bank-additional-full.csv' in expected directories.")

print(f"Loading raw dataset from: {raw_data_path}")
df_raw = pd.read_csv(raw_data_path, sep=';')
print(f"Initial Raw Dataset Dimensions: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns")

# ADR-003 & EDA-17 Deduplication Policy: Prune exact duplicate rows prior to splitting
n_dups = df_raw.duplicated().sum()
df_raw.drop_duplicates(inplace=True)
print(f"Deduplicated Dataset Dimensions: {df_raw.shape[0]:,} rows × {df_raw.shape[1]} columns ({n_dups} duplicate rows pruned, preserving 41,176 records)")
assert len(df_raw) == 41176, f"Expected 41,176 records after deduplication, got {len(df_raw)}"

print("\nFirst 3 records:")
print(df_raw.head(3))

Loading raw dataset from: /Users/moni/Desktop/ML proj/data/raw/bank-additional-full.csv
Initial Raw Dataset Dimensions: 41,188 rows × 21 columns
Deduplicated Dataset Dimensions: 41,176 rows × 21 columns (12 duplicate rows pruned, preserving 41,176 records)

First 3 records:
   age        job  marital    education  default housing loan    contact month day_of_week  duration  campaign  pdays  previous     poutcome  emp.var.rate  cons.price.idx  cons.conf.idx  euribor3m  nr.employed   y
0   56  housemaid  married     basic.4y       no      no   no  telephone   may         mon       261         1    999         0  nonexistent           1.1          93.994          -36.4      4.857       5191.0  no
1   57   services  married  high.school  unknown      no   no  telephone   may         mon       149         1    999         0  nonexistent           1.1          93.994          -36.4      4.857       5191.0  no
2   37   services  married  high.school       no     yes   no  telephone   may     

In [3]:
# 1. Target encoding: 'yes' -> 1, 'no' -> 0
target_col = 'y'
y = df_raw[target_col].map({'no': 0, 'yes': 1}).astype(int)

# 2. PRE-01 / ADR-003: Drop target and post-event leakage variable 'duration'
leakage_col = 'duration'
X = df_raw.drop(columns=[target_col, leakage_col])

print(f"Target variable '{target_col}' encoded as binary (0/1):")
print(y.value_counts(normalize=True).rename('proportion').to_frame().assign(count=y.value_counts()))

print(f"\nFeature Matrix X shape after discarding '{leakage_col}' (PRE-01): {X.shape}")
print(f"Is '{leakage_col}' present in X? {'duration' in X.columns}")

Target variable 'y' encoded as binary (0/1):
   proportion  count
y                   
0    0.887337  36537
1    0.112663   4639

Feature Matrix X shape after discarding 'duration' (PRE-01): (41176, 19)
Is 'duration' present in X? False


## 2. Validation Split Strategy Investigation & Rationale

### Scope Discrepancy Context
The initial project submission proposed a validation methodology that considers the chronological ordering of the campaign data to avoid leaking time-related patterns into holdout sets. However, before establishing our production preprocessing pipeline, we conducted an empirical investigation comparing:
- **Option A – Temporal Holdout**: Partitioning the dataset chronologically (first 80% development, last 20% holdout).
- **Option B – Stratified Random Holdout**: Partitioning via stratified random sampling (80% development, 20% holdout with `random_state=42`).

The comparison below details why Option B is retained and formally ratified in `docs/decision_log.md` (ADR-001).

In [4]:
# Empirical Investigation: Option A (Temporal Split) vs Option B (Stratified Random Split)

# Option A: Chronological 80/20 split
split_idx = int(len(df_raw) * 0.8)
train_chrono = df_raw.iloc[:split_idx]
test_chrono = df_raw.iloc[split_idx:]

y_chrono_train = (train_chrono['y'] == 'yes').astype(int)
y_chrono_test = (test_chrono['y'] == 'yes').astype(int)

# Option B: Stratified 80/20 split (for comparison)
X_dummy_tr, X_dummy_te, y_strat_tr, y_strat_te = train_test_split(
    X, y, test_size=0.20, random_state=RANDOM_STATE, stratify=y
)

print("=" * 80)
print("EMPIRICAL COMPARISON: TEMPORAL VS STRATIFIED RANDOM SPLIT (80/20)")
print("=" * 80)

# 1. Target class distribution comparison
print("\n1. TARGET CLASS BASE RATE (PRIOR):")
print(f" - Full Dataset Positives:     {y.mean():.4%} ({y.sum():,} / {len(y):,})")
print(f" - Option A (Temporal) Train:  {y_chrono_train.mean():.4%} ({y_chrono_train.sum():,} / {len(y_chrono_train):,})")
print(f" - Option A (Temporal) Test:   {y_chrono_test.mean():.4%} ({y_chrono_test.sum():,} / {len(y_chrono_test):,})")
print(f"   --> Relative Prior Shift:   +{(y_chrono_test.mean() / y_chrono_train.mean() - 1):.1%}")
print(f" - Option B (Stratified) Train:{y_strat_tr.mean():.4%} ({y_strat_tr.sum():,} / {len(y_strat_tr):,})")
print(f" - Option B (Stratified) Test: {y_strat_te.mean():.4%} ({y_strat_te.sum():,} / {len(y_strat_te):,})")

# 2. Covariate Shift Comparison (Macroeconomic Indicators)
econ_vars = ['euribor3m', 'nr.employed', 'emp.var.rate']
print("\n2. MACROECONOMIC COVARIATE SHIFT (OPTION A TEMPORAL):")
for col in econ_vars:
    tr_min, tr_max, tr_mean = train_chrono[col].min(), train_chrono[col].max(), train_chrono[col].mean()
    te_min, te_max, te_mean = test_chrono[col].min(), test_chrono[col].max(), test_chrono[col].mean()
    print(f" - {col:14s} | Train: [{tr_min:.3f}, {tr_max:.3f}] (mean {tr_mean:.3f}) | Test: [{te_min:.3f}, {te_max:.3f}] (mean {te_mean:.3f})")

# 3. Categorical Support Coverage
train_months = set(train_chrono['month'].unique())
test_months = set(test_chrono['month'].unique())
missing_in_train = test_months - train_months
print(f"\n3. CATEGORICAL SUPPORT:")
print(f" - Months in Temporal Train: {sorted(list(train_months))}")
print(f" - Months in Temporal Test:  {sorted(list(test_months))}")
if missing_in_train:
    print(f" - WARNING: Month category missing from training set under Option A: {missing_in_train}")

print("=" * 80)

EMPIRICAL COMPARISON: TEMPORAL VS STRATIFIED RANDOM SPLIT (80/20)

1. TARGET CLASS BASE RATE (PRIOR):
 - Full Dataset Positives:     11.2663% (4,639 / 41,176)
 - Option A (Temporal) Train:  6.3752% (2,100 / 32,940)
 - Option A (Temporal) Test:   30.8281% (2,539 / 8,236)
   --> Relative Prior Shift:   +383.6%
 - Option B (Stratified) Train:11.2659% (3,711 / 32,940)
 - Option B (Stratified) Test: 11.2676% (928 / 8,236)

2. MACROECONOMIC COVARIATE SHIFT (OPTION A TEMPORAL):
 - euribor3m      | Train: [1.299, 5.045] (mean 4.269) | Test: [0.634, 1.299] (mean 1.031)
 - nr.employed    | Train: [5099.100, 5228.100] (mean 5195.140) | Test: [4963.600, 5099.100] (mean 5054.628)
 - emp.var.rate   | Train: [-1.800, 1.400] (mean 0.650) | Test: [-3.400, -1.100] (mean -2.192)

3. CATEGORICAL SUPPORT:
 - Months in Temporal Train: ['apr', 'aug', 'dec', 'jul', 'jun', 'mar', 'may', 'nov', 'oct']
 - Months in Temporal Test:  ['apr', 'aug', 'dec', 'jul', 'jun', 'mar', 'may', 'nov', 'oct', 'sep']
 - WARNING:

### Summary of Findings & Decision (ADR-001 & ADR-003)
As demonstrated by the empirical comparison above and ratified under **[ADR-001](file:///Users/moni/Desktop/ML%20proj/docs/decision_log.md#adr-001)**:
- **Option A (Temporal Split)** introduces severe macroeconomic regime shifts ($euribor3m$ zero support overlap) and an extreme $+384\%$ shift in positive base rate ($6.37\%$ to $30.83\%$), rendering model cross-validation and probability calibration unviable.
- **Option B (Stratified Random Split)** is ratified as the primary partitioning strategy with explicit seed `random_state=42`. Combining Option B with **[ADR-003 Deduplication](file:///Users/moni/Desktop/ML%20proj/docs/decision_log.md#adr-003)** ensures identical $11.27\%$ priors, uniform covariate coverage, and zero duplicate leakage across partitions.

In [5]:
# Ratified Execution: Stratified 80/20 train/test split with contract verification (ADR-001 & ADR-003)
X_train, X_test, y_train, y_test = train_test_split(
    X, y,
    test_size=0.20,
    random_state=RANDOM_STATE,
    stratify=y
)

# Contract Invariant Assertions (ADR-001 & ADR-003 Reconciled: 41,176 records)
assert len(X_train) == 32940, f"Expected 32,940 train records, got {len(X_train)}"
assert len(X_test) == 8236, f"Expected 8,236 test records, got {len(X_test)}"
assert len(y_train) == 32940
assert len(y_test) == 8236
assert len(X_train) + len(X_test) == 41176

# Mutual Exclusivity Assertion
train_idx_set = set(X_train.index)
test_idx_set = set(X_test.index)
assert len(train_idx_set.intersection(test_idx_set)) == 0, "Index overlap detected between train and test sets!"

# Cross-Split Duplicate Profile Leakage Assertion (Zero duplicate profiles across partitions)
dup_profiles = df_raw[df_raw.duplicated(keep=False)].drop(columns=['y'])
if len(dup_profiles) > 0:
    cross_split_leak = pd.merge(
        X_train.loc[X_train.index.isin(dup_profiles.index)].reset_index(),
        X_test.loc[X_test.index.isin(dup_profiles.index)].reset_index(),
        on=list(X.columns)
    )
    assert len(cross_split_leak) == 0, f"Cross-split duplicate leakage detected: {len(cross_split_leak)} rows!"

# Stratification Balance Assertion (< 0.1% delta)
strat_delta = abs(y_train.mean() - y_test.mean())
assert strat_delta < 0.001, f"Stratification mismatch: delta={strat_delta:.6f}"

print("Data Split Summary (Ratified Option B with ADR-003 Deduplication):")
print(f" - Development Set (X_train): {X_train.shape[0]:,} records ({X_train.shape[0]/len(X):.1%})")
print(f" - Holdout Test Set (X_test):   {X_test.shape[0]:,} records ({X_test.shape[0]/len(X):.1%})")
print(f"\nTarget Class Balance Verification:")
print(f" - Full Dataset Positives:      {y.mean():.4%} ({y.sum():,} / {len(y):,})")
print(f" - Train Set Positives:         {y_train.mean():.4%} ({y_train.sum():,} / {len(y_train):,})")
print(f" - Test Set Positives:          {y_test.mean():.4%} ({y_test.sum():,} / {len(y_test):,})")
print(f" - Stratification Delta:        {strat_delta:.6f}")
print("Stratification successfully verified, zero duplicate leakage, and contract invariants passed.")

Data Split Summary (Ratified Option B with ADR-003 Deduplication):
 - Development Set (X_train): 32,940 records (80.0%)
 - Holdout Test Set (X_test):   8,236 records (20.0%)

Target Class Balance Verification:
 - Full Dataset Positives:      11.2663% (4,639 / 41,176)
 - Train Set Positives:         11.2659% (3,711 / 32,940)
 - Test Set Positives:          11.2676% (928 / 8,236)
 - Stratification Delta:        0.000017
Stratification successfully verified, zero duplicate leakage, and contract invariants passed.


## 3. Modular Feature Engineering (`src.transformers`)

In adherence to **Constitution Principle II (*Pipeline-First*)**, domain feature transformations are encapsulated in [`src/transformers.py`](../src/transformers.py) via `BankFeatureEngineer`:
- **PRE-02**: Treating sentinel `pdays = 999` ($96.32\%$ uncontacted) as a discrete behavioral state rather than a continuous metric distance.
- **PRE-07**: Discretizing elapsed days into non-linear recency cohorts (`pdays_group`: `0_to_6_days`, `7_to_14_days`, `15_plus_days`, `not_contacted`).
- **PRE-08**: Deriving binary interaction indicator `previously_contacted` ($pdays \ne 999$).
- **PRE-09 & ADR-004**: Life-stage non-linearity engineering via categorical `age_group` (`<30`, `30-39`, `40-49`, `50-59`, `60+`) while intentionally preserving continuous standardized `age` (dual representation).
- **PRE-10**: Dropping raw `pdays` column to eliminate sentinel 999 from distance and scale computations.
- **ADR-003**: Preserving historical campaign features (`previous` count and `poutcome` status) to retain information regarding the 4,110 legacy CRM failure contacts.

In [6]:
# Ensure project root is in sys.path for modular imports
import sys
from pathlib import Path

# Deterministically ensure PROJECT_ROOT is defined before use
if 'PROJECT_ROOT' not in globals():
    current_dir = Path.cwd().resolve()
    PROJECT_ROOT = next(
        (p for p in [current_dir] + list(current_dir.parents) if (p / 'src' / 'transformers.py').exists()),
        current_dir
    )

for p in ['.', '..', str(PROJECT_ROOT)]:
    if (Path(p) / 'src').exists() and str(p) not in sys.path:
        sys.path.insert(0, str(p))

from src.transformers import BankFeatureEngineer

# Test feature engineer transformer on X_train only
fe_test = BankFeatureEngineer()
X_train_fe_sample = fe_test.transform(X_train)

print("Engineered feature verification:")
print(" - New columns added:", [c for c in X_train_fe_sample.columns if c not in X_train.columns])
print(" - Columns removed:   ", [c for c in X_train.columns if c not in X_train_fe_sample.columns])
print("\n'pdays_group' breakdown:")
print(X_train_fe_sample['pdays_group'].value_counts())
print("\n'age_group' breakdown:")
print(X_train_fe_sample['age_group'].value_counts())

Engineered feature verification:
 - New columns added: ['previously_contacted', 'pdays_group', 'age_group']
 - Columns removed:    ['pdays']

'pdays_group' breakdown:
pdays_group
not_contacted    31733
0_to_6_days        881
7_to_14_days       272
15_plus_days        54
Name: count, dtype: int64

'age_group' breakdown:
age_group
30-39    13507
40-49     8426
50-59     5440
<30       4590
60+        977
Name: count, dtype: int64


## 4. Categorical Encoding Strategy: Education Feature Representation (ADR-002, PRE-05)

Initial data engineering heuristic treated `education` as an ordinal feature (`unknown`=0, `illiterate`=1, ..., `university.degree`=7). As evaluated in **[ADR-002](file:///Users/moni/Desktop/ML%20proj/docs/decision_log.md#adr-002)** and **[PRE-05](file:///Users/moni/Desktop/ML%20proj/docs/preprocessing_feature_log.md)**, this was defective because:
1. It conflated missingness (`unknown`) with lowest attainment;
2. It violated the non-monotonic empirical conversion profile (`illiterate`: $22.2\%$, `basic.9y`: $7.8\%$, `univ`: $13.7\%$, `unknown`: $14.5\%$).

In [7]:
# Empirical Investigation: Categorical Encoding Strategies for Education
from scipy.stats import chi2_contingency
from sklearn.linear_model import LogisticRegression
from sklearn.ensemble import RandomForestClassifier, HistGradientBoostingClassifier
from sklearn.model_selection import StratifiedKFold, cross_validate
from sklearn.metrics import roc_auc_score, average_precision_score, log_loss

print("=" * 80)
print("EMPIRICAL COMPARISON: EDUCATION ENCODING STRATEGIES")
print("=" * 80)

# 1. Demographic Distribution & Empirical Response Rates
edu_stats = df_raw.groupby('education')['y'].agg(
    total='count',
    positives=lambda s: (s == 'yes').sum(),
    response_rate=lambda s: (s == 'yes').mean()
)
edu_stats['pct_total'] = edu_stats['total'] / len(df_raw) * 100
edu_stats['response_rate_pct'] = edu_stats['response_rate'] * 100
order = ['unknown', 'illiterate', 'basic.4y', 'basic.6y', 'basic.9y', 'high.school', 'professional.course', 'university.degree']
edu_stats = edu_stats.loc[order]

print("\n1. EDUCATION CATEGORY DISTRIBUTION & CONVERSION RATES:")
for idx, r in edu_stats.iterrows():
    print(f" - {idx:20s} | N={r.total:5.0f} ({r.pct_total:5.2f}%) | Responders: {r.positives:4.0f} | Rate: {r.response_rate_pct:5.2f}%")

chi2, p_val, dof, _ = chi2_contingency(pd.crosstab(df_raw['education'], df_raw['y']))
print(f"\nChi-Square Test of Independence: chi2={chi2:.2f}, p={p_val:.2e} (df={dof})")

# 2. Univariate Predictive Capacity (Ordinal vs OHE)
y_all = (df_raw['y'] == 'yes').astype(int)
X_ord_uni = OrdinalEncoder(categories=[order]).fit_transform(df_raw[['education']])
lr_ord_uni = LogisticRegression().fit(X_ord_uni, y_all)
prob_ord_uni = lr_ord_uni.predict_proba(X_ord_uni)[:, 1]

X_ohe_uni = OneHotEncoder().fit_transform(df_raw[['education']])
lr_ohe_uni = LogisticRegression().fit(X_ohe_uni, y_all)
prob_ohe_uni = lr_ohe_uni.predict_proba(X_ohe_uni)[:, 1]

print("\n2. UNIVARIATE PREDICTIVE POWER:")
print(f" - Option A (Current Ordinal):  ROC-AUC: {roc_auc_score(y_all, prob_ord_uni):.4f} | Log-Loss: {log_loss(y_all, prob_ord_uni):.5f}")
print(f" - Option B (One-Hot Encoding): ROC-AUC: {roc_auc_score(y_all, prob_ohe_uni):.4f} | Log-Loss: {log_loss(y_all, prob_ohe_uni):.5f}")

# 3. Multivariate Model Comparison Across Pipeline Candidates
nominal_cols_base = [
    'job', 'marital', 'default', 'housing', 'loan',
    'contact', 'month', 'day_of_week', 'poutcome',
    'age_group', 'pdays_group'
]
robust_num_cols = ['campaign', 'previous']
standard_num_cols = ['age', 'emp.var.rate', 'cons.price.idx', 'cons.conf.idx', 'euribor3m', 'nr.employed']
passthrough_cols = ['previously_contacted']

# Pipeline A: Current Ordinal
prep_strat_a = ColumnTransformer([
    ('nominal', OneHotEncoder(handle_unknown='ignore', sparse_output=False), nominal_cols_base),
    ('ordinal', OrdinalEncoder(categories=[order], handle_unknown='use_encoded_value', unknown_value=-1), ['education']),
    ('robust_num', RobustScaler(), robust_num_cols),
    ('standard_num', StandardScaler(), standard_num_cols),
    ('passthrough', 'passthrough', passthrough_cols)
])
pipe_strat_a = Pipeline([('fe', BankFeatureEngineer()), ('prep', prep_strat_a)])

# Pipeline B: One-Hot Encoding
prep_strat_b = ColumnTransformer([
    ('nominal', OneHotEncoder(handle_unknown='ignore', sparse_output=False), nominal_cols_base + ['education']),
    ('robust_num', RobustScaler(), robust_num_cols),
    ('standard_num', StandardScaler(), standard_num_cols),
    ('passthrough', 'passthrough', passthrough_cols)
])
pipe_strat_b = Pipeline([('fe', BankFeatureEngineer()), ('prep', prep_strat_b)])

X_tr_a = pipe_strat_a.fit_transform(X_train)
X_te_a = pipe_strat_a.transform(X_test)

X_tr_b = pipe_strat_b.fit_transform(X_train)
X_te_b = pipe_strat_b.transform(X_test)

# Compare Logistic Regression Coefficients on Education
lr_a = LogisticRegression(max_iter=1000, random_state=42).fit(X_tr_a, y_train)
lr_b = LogisticRegression(max_iter=1000, random_state=42).fit(X_tr_b, y_train)

ord_feat_idx = [i for i, c in enumerate(pipe_strat_a.named_steps['prep'].get_feature_names_out()) if 'education' in c][0]
ohe_feat_indices = [i for i, c in enumerate(pipe_strat_b.named_steps['prep'].get_feature_names_out()) if 'education' in c]
ohe_feat_names = [pipe_strat_b.named_steps['prep'].get_feature_names_out()[i] for i in ohe_feat_indices]

print("\n3. LOGISTIC REGRESSION WEIGHT CONSTRAINTS:")
print(f" - Ordinal Education Learned Slope: {lr_a.coef_[0][ord_feat_idx]:.5f} (forces single monotonic slope across all levels)")
print(" - One-Hot Education Learned Weights (independent per category):")
for name, idx in zip(ohe_feat_names, ohe_feat_indices):
    cat_name = name.replace('nominal__education_', '')
    print(f"   * {cat_name:20s}: {lr_b.coef_[0][idx]:+.5f}")

# 4. Cross-Validation and Holdout Benchmark Across Model Families
cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=RANDOM_STATE)
bench_models = {
    'LogisticRegression': LogisticRegression(max_iter=1000, random_state=42),
    'RandomForest': RandomForestClassifier(n_estimators=100, max_depth=10, random_state=42, n_jobs=-1),
    'HistGradientBoosting': HistGradientBoostingClassifier(max_iter=100, random_state=42)
}

print("\n4. 5-FOLD STRATIFIED CV & HOLDOUT BENCHMARKS:")
for s_label, X_tr_m, X_te_m in [("Option A (Current Ordinal - 64 features)", X_tr_a, X_te_a), ("Option B (One-Hot Encoding - 71 features)", X_tr_b, X_te_b)]:
    print(f"\n[{s_label}]")
    for m_label, model_obj in bench_models.items():
        cv_scores = cross_validate(model_obj, X_tr_m, y_train, cv=cv, scoring=['roc_auc', 'average_precision'], n_jobs=-1)
        model_obj.fit(X_tr_m, y_train)
        y_test_pred = model_obj.predict_proba(X_te_m)[:, 1]
        t_auc = roc_auc_score(y_test, y_test_pred)
        t_ap = average_precision_score(y_test, y_test_pred)
        print(f" - {m_label:22s} | CV ROC-AUC: {cv_scores['test_roc_auc'].mean():.4f} +/- {cv_scores['test_roc_auc'].std():.4f} | CV PR-AUC: {cv_scores['test_average_precision'].mean():.4f} | Test ROC-AUC: {t_auc:.4f} | Test PR-AUC: {t_ap:.4f}")

print("=" * 80)

EMPIRICAL COMPARISON: EDUCATION ENCODING STRATEGIES

1. EDUCATION CATEGORY DISTRIBUTION & CONVERSION RATES:
 - unknown              | N= 1730 ( 4.20%) | Responders:  251 | Rate: 14.51%
 - illiterate           | N=   18 ( 0.04%) | Responders:    4 | Rate: 22.22%
 - basic.4y             | N= 4176 (10.14%) | Responders:  428 | Rate: 10.25%
 - basic.6y             | N= 2291 ( 5.56%) | Responders:  188 | Rate:  8.21%
 - basic.9y             | N= 6045 (14.68%) | Responders:  473 | Rate:  7.82%
 - high.school          | N= 9512 (23.10%) | Responders: 1031 | Rate: 10.84%
 - professional.course  | N= 5240 (12.73%) | Responders:  595 | Rate: 11.35%
 - university.degree    | N=12164 (29.54%) | Responders: 1669 | Rate: 13.72%

Chi-Square Test of Independence: chi2=192.85, p=3.75e-38 (df=7)

2. UNIVARIATE PREDICTIVE POWER:
 - Option A (Current Ordinal):  ROC-AUC: 0.5406 | Log-Loss: 0.35153
 - Option B (One-Hot Encoding): ROC-AUC: 0.5598 | Log-Loss: 0.34967

3. LOGISTIC REGRESSION WEIGHT CONSTRAINTS

### Summary of Findings & Decision (ADR-002 & PRE-05)

As ratified under **[ADR-002](file:///Users/moni/Desktop/ML%20proj/docs/decision_log.md#adr-002)**, the arbitrary ordinal hierarchy for `education` is completely replaced by **One-Hot Encoding (Option B)** across all 8 attainment levels (including `unknown` as an orthogonal state). This yields $+191$ bps univariate ROC-AUC lift and unconstrained coefficient estimation across linear and tree models.

## 4.5 Project-Wide Policy for "Unknown" Categorical States (PRE-04, ADR-003)

**Ratified Handling Policy (PRE-04 & ADR-003)**:
- **No rows are silently dropped** because of `"unknown"`. Dropping records would discard $>26\%$ of the dataset.
- `"unknown"` is explicitly retained as a separate valid category across all nominal variables (`job`, `marital`, `default`, `housing`, `loan`, `education`). Missingness in CRM data is informative (e.g. `default='unknown'` converts at $5.15\%$ vs $12.88\%$ for clean credit).
- All nominal attributes are passed through `OneHotEncoder(handle_unknown='ignore', sparse_output=False)`.

## 5. End-to-End Scikit-Learn Preprocessing Pipeline Architecture (PRE-01..10, ADR-001..004)

The production preprocessing pipeline integrates all ratified architectural decisions into a single Scikit-Learn `Pipeline`:
- **PRE-01**: `duration` quarantined via `remainder='drop'` (zero post-event target leakage);
- **PRE-02, PRE-07, PRE-08, PRE-09, PRE-10**: Engineered domain features generated via `BankFeatureEngineer`;
- **PRE-03**: Mixed scaling applying `RobustScaler` to heavy-tailed counts (`campaign`, `previous`) and `StandardScaler` to macroeconomic variables and continuous `age`;
- **PRE-04, PRE-05, PRE-06, PRE-09**: Nominal categorical features encoded via `OneHotEncoder(handle_unknown='ignore', sparse_output=False)` producing orthogonal binary indicator columns;
- Total pipeline output dimensionality: **71 orthogonal features**.

In [8]:
# 1. Feature column definitions
# PRE-04, PRE-05, PRE-06, PRE-07, PRE-09: Nominal One-Hot Encoded columns
nominal_cols = [
    'job', 'marital', 'default', 'housing', 'loan',
    'contact', 'month', 'day_of_week', 'poutcome',
    'education',       # PRE-05 / ADR-002: 8 binary indicator columns
    'age_group',       # PRE-09 / ADR-004: 5 life-stage cohorts (<30, 30-39, 40-49, 50-59, 60+)
    'pdays_group'      # PRE-07: 4 recency cohorts (0_to_6_days, 7_to_14_days, 15_plus_days, not_contacted)
]

# PRE-03: Heavy-tail robust scaling for count variables with extreme outliers
robust_num_cols = ['campaign', 'previous']

# PRE-03 & PRE-09 (ADR-004): Standardization for macroeconomic indicators and continuous age
standard_num_cols = [
    'age',             # PRE-09: Continuous standardized age preserved for fine-grained ranking
    'emp.var.rate', 'cons.price.idx',
    'cons.conf.idx', 'euribor3m', 'nr.employed'
]

# PRE-08: Binary prior contact flag passthrough
passthrough_cols = ['previously_contacted']

# 2. Build ColumnTransformer (PRE-01, PRE-03, PRE-04, PRE-05, PRE-06, PRE-09)
preprocessor = ColumnTransformer(
    transformers=[
        ('nominal', OneHotEncoder(handle_unknown='ignore', sparse_output=False), nominal_cols),
        ('robust_num', RobustScaler(), robust_num_cols),
        ('standard_num', StandardScaler(), standard_num_cols),
        ('passthrough', 'passthrough', passthrough_cols)
    ],
    remainder='drop',  # PRE-01: Quarantines post-event duration leakage
    verbose_feature_names_out=False
)

# 3. Assemble full preprocessing pipeline (PRE-01..PRE-10, ADR-001..ADR-004)
full_pipeline = Pipeline([
    ('feature_engineer', BankFeatureEngineer(validate_input=True, flag_inconsistencies=False)),
    ('preprocessor', preprocessor)
])

print("Full Pipeline Architecture (Ratified ADR-001..004 & PRE-01..10):")
print(full_pipeline)

Full Pipeline Architecture (Ratified ADR-001..004 & PRE-01..10):
Pipeline(steps=[('feature_engineer', BankFeatureEngineer()),
                ('preprocessor',
                 ColumnTransformer(transformers=[('nominal',
                                                  OneHotEncoder(handle_unknown='ignore',
                                                                sparse_output=False),
                                                  ['job', 'marital', 'default',
                                                   'housing', 'loan', 'contact',
                                                   'month', 'day_of_week',
                                                   'poutcome', 'education',
                                                   'age_group',
                                                   'pdays_group']),
                                                 ('robust_num', RobustScaler(),
                                                  ['campaign', 'previous']),
    

## 6. Strict Fit-on-Train Pipeline Execution (Zero Data Leakage)
In compliance with Constitution Principle III (*No Data Leakage*), all transformer scaling parameters ($\mu, \sigma$), categorical one-hot dictionaries, and grouping statistics are fitted strictly on `X_train`. The holdout set `X_test` is transformed using the fitted pipeline without parameter re-estimation.

In [9]:
# Fit on training data ONLY
print("Fitting preprocessing pipeline on training data...")
full_pipeline.fit(X_train)

# Transform train and holdout test sets
X_train_proc = full_pipeline.transform(X_train)
X_test_proc = full_pipeline.transform(X_test)

# Extract column names from fitted ColumnTransformer
feature_names = full_pipeline.named_steps['preprocessor'].get_feature_names_out()

# Create clean DataFrames
df_X_train_proc = pd.DataFrame(X_train_proc, columns=feature_names, index=X_train.index)
df_X_test_proc = pd.DataFrame(X_test_proc, columns=feature_names, index=X_test.index)

print("\nPreprocessing Transformation Completed Successfully:")
print(f" - Transformed Training Set: {df_X_train_proc.shape[0]:,} rows × {df_X_train_proc.shape[1]} features")
print(f" - Transformed Test Set:     {df_X_test_proc.shape[0]:,} rows × {df_X_test_proc.shape[1]} features")
print(f" - Total Missing / NaN Values in Train: {df_X_train_proc.isna().sum().sum()}")
print(f" - Total Missing / NaN Values in Test:  {df_X_test_proc.isna().sum().sum()}")

# Contract assertions (ADR-002: 71 features; ADR-003: 32,940 train / 8,236 test)
assert df_X_train_proc.shape == (32940, 71), f"Expected (32940, 71), got {df_X_train_proc.shape}"
assert df_X_test_proc.shape == (8236, 71), f"Expected (8236, 71), got {df_X_test_proc.shape}"
assert df_X_train_proc.isna().sum().sum() == 0, "NaN values detected in training set!"
assert df_X_test_proc.isna().sum().sum() == 0, "NaN values detected in test set!"
assert 'duration' not in feature_names, "CRITICAL: 'duration' found in feature names!"
assert 'education_unknown' in feature_names, "CRITICAL: 'education_unknown' missing from feature names!"
assert 'age_group_<30' in feature_names, "CRITICAL: 'age_group_<30' missing from feature names!"
assert 'age' in feature_names, "CRITICAL: 'age' missing from feature names!"
assert 'previously_contacted' in feature_names, "CRITICAL: 'previously_contacted' missing from feature names!"
print("Integrity checks passed: 0 missing values, 71 features, zero leakage, and ADR-001..004 verified.")

Fitting preprocessing pipeline on training data...

Preprocessing Transformation Completed Successfully:
 - Transformed Training Set: 32,940 rows × 71 features
 - Transformed Test Set:     8,236 rows × 71 features
 - Total Missing / NaN Values in Train: 0
 - Total Missing / NaN Values in Test:  0
Integrity checks passed: 0 missing values, 71 features, zero leakage, and ADR-001..004 verified.


## 7. Post-Scaling Feature Distributions Inspection

In [10]:
print("Sample of Processed Feature Matrix (First 3 rows, selected columns including ADR-002 features):")
selected_sample_cols = [
    'education_university.degree', 'education_unknown', 'previously_contacted',
    'pdays_group_0_to_6_days', 'pdays_group_not_contacted',
    'age_group_<30', 'age_group_60+', 'campaign', 'euribor3m', 'nr.employed'
]
print(df_X_train_proc[selected_sample_cols].head(3))

# Diagnostic Plots
fig, axes = plt.subplots(1, 3, figsize=(16, 5))

# 1. Age distribution standardized
sns.histplot(df_X_train_proc['age'], ax=axes[0], kde=True, color='#2b5c8f', bins=30)
axes[0].set_title("Standardized Age (Mean=0, Std=1)")
axes[0].set_xlabel("Z-Score")

# 2. Campaign distribution robust-scaled
sns.histplot(df_X_train_proc['campaign'], ax=axes[1], kde=False, color='#e07a5f', bins=30)
axes[1].set_title("Robust-Scaled Campaign (Median=0, IQR=1)")
axes[1].set_xlabel("Robust Scaled Value")
axes[1].set_xlim(-1, 10)

# 3. Euribor 3-month rate standardized
sns.histplot(df_X_train_proc['euribor3m'], ax=axes[2], kde=True, color='#81b29a', bins=30)
axes[2].set_title("Standardized Euribor 3M Rate")
axes[2].set_xlabel("Z-Score")

plt.suptitle("Post-Scaling Feature Distributions", fontsize=14, weight='bold')
plt.close(fig)

Sample of Processed Feature Matrix (First 3 rows, selected columns including ADR-002 features):
       education_university.degree  education_unknown  previously_contacted  pdays_group_0_to_6_days  pdays_group_not_contacted  age_group_<30  age_group_60+  campaign  euribor3m  nr.employed
36454                          0.0                0.0                   0.0                      0.0                        1.0            1.0            0.0       0.0  -1.357248    -1.252900
1233                           0.0                0.0                   0.0                      0.0                        1.0            0.0            0.0      -0.5   0.712567     0.332805
24111                          1.0                0.0                   0.0                      0.0                        1.0            0.0            0.0      -0.5   0.361165     0.399107


## 8. Exporting Processed Dataset Artifacts & Pipeline Serialization
All processed datasets are saved to `data/processed/`, accompanied by the fitted pipeline object and metadata schema reflecting ADR-001 and ADR-002.

In [11]:
# Create output directory anchored to PROJECT_ROOT
processed_dir = (PROJECT_ROOT / "data" / "processed") if 'PROJECT_ROOT' in globals() else (Path("data/processed") if Path("data").exists() else Path("../data/processed"))
processed_dir.mkdir(parents=True, exist_ok=True)

# 1. Save processed feature sets
train_features_path = processed_dir / "X_train_processed.csv"
test_features_path = processed_dir / "X_test_processed.csv"
train_target_path = processed_dir / "y_train.csv"
test_target_path = processed_dir / "y_test.csv"

df_X_train_proc.to_csv(train_features_path, index=False)
df_X_test_proc.to_csv(test_features_path, index=False)
y_train.to_csv(train_target_path, index=False)
y_test.to_csv(test_target_path, index=False)

# 2. Serialize full fitted pipeline
pipeline_save_path = processed_dir / "preprocessor_pipeline.joblib"
joblib.dump(full_pipeline, pipeline_save_path)

# 3. Export feature catalog & metadata schema (ADR-001..ADR-004, PRE-01..PRE-10)
metadata = {
    "project": "UCI Bank Marketing Campaign Response",
    "phase": "02_Data_Preprocessing_Pipeline",
    "author": "Athukorala M. B. (IT24101937)",
    "git_branch": "005-sync-notebooks-with-logs",
    "validation_split_strategy": "Option B - Stratified Random Holdout (80/20, seed=42)",
    "deduplication_policy": "Exact duplicate rows pruned prior to splitting (ADR-003, EDA-17)",
    "education_encoding_strategy": "Option B - One-Hot Encoding (including unknown as distinct category, ADR-002)",
    "age_representation_strategy": "Dual Representation - Retaining both continuous 'age' and categorical 'age_group' (PRE-09, ADR-004)",
    "decision_log_reference": "docs/decision_log.md#adr-001, docs/decision_log.md#adr-002, docs/decision_log.md#adr-003, docs/decision_log.md#adr-004",
    "total_input_features": X.shape[1],
    "total_output_features": len(feature_names),
    "train_records": len(df_X_train_proc),
    "test_records": len(df_X_test_proc),
    "train_positive_rate": float(y_train.mean()),
    "test_positive_rate": float(y_test.mean()),
    "leakage_feature_dropped": "duration",
    "features_list": list(feature_names)
}

metadata_path = processed_dir / "feature_metadata.json"
with open(metadata_path, 'w') as f:
    json.dump(metadata, f, indent=2)

print("Export completed successfully:")
print(f" - X_train:  {train_features_path} ({train_features_path.stat().st_size / 1024 / 1024:.2f} MB)")
print(f" - X_test:   {test_features_path} ({test_features_path.stat().st_size / 1024 / 1024:.2f} MB)")
print(f" - y_train:  {train_target_path}")
print(f" - y_test:   {test_target_path}")
print(f" - Pipeline: {pipeline_save_path}")
print(f" - Metadata: {metadata_path}")

Export completed successfully:
 - X_train:  /Users/moni/Desktop/ML proj/data/processed/X_train_processed.csv (11.86 MB)
 - X_test:   /Users/moni/Desktop/ML proj/data/processed/X_test_processed.csv (2.97 MB)
 - y_train:  /Users/moni/Desktop/ML proj/data/processed/y_train.csv
 - y_test:   /Users/moni/Desktop/ML proj/data/processed/y_test.csv
 - Pipeline: /Users/moni/Desktop/ML proj/data/processed/preprocessor_pipeline.joblib
 - Metadata: /Users/moni/Desktop/ML proj/data/processed/feature_metadata.json


## 9. Serialization & Inference Verification

In [12]:
# Reload serialized pipeline
loaded_pipeline = joblib.load(pipeline_save_path)

# Verify estimator parameters inspection (contract requirement)
params = loaded_pipeline.named_steps['feature_engineer'].get_params()
assert 'validate_input' in params, "validate_input missing from parameters"
assert 'flag_inconsistencies' in params, "flag_inconsistencies missing from parameters"

# Take 5 raw samples from X_test (before any transformation)
raw_samples = X_test.head(5)

# Transform using loaded pipeline
transformed_samples = loaded_pipeline.transform(raw_samples)

print(f"Loaded Pipeline Type: {type(loaded_pipeline)}")
print(f"Raw Input Shape:      {raw_samples.shape}")
print(f"Transformed Output:   {transformed_samples.shape}")

# Verify exact numerical match with original transformed test set
original_subset = df_X_test_proc.head(5).to_numpy()
np.testing.assert_allclose(transformed_samples, original_subset, rtol=1e-5, atol=1e-8)
print("Verification Passed: Loaded pipeline transforms raw incoming leads identically into 71 features with zero attribute errors!")

Loaded Pipeline Type: <class 'sklearn.pipeline.Pipeline'>
Raw Input Shape:      (5, 19)
Transformed Output:   (5, 71)
Verification Passed: Loaded pipeline transforms raw incoming leads identically into 71 features with zero attribute errors!
